### PR TITLE
Update Help Command

### PR DESCRIPTION
**Pull request by exsersewo**

### CODE REVIEW

**Comment by velddev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 19
           

Complexity increasing per file
==============================
- src/Miki/Services/CommandTreeService.cs  2
- src/Miki/Modules/GeneralModule.cs  1
- src/Miki/Attributes/EmojiAttribute.cs  1
         

Clones added
============
- src/Miki/Services/CommandTreeService.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/velddev/Miki/pullRequest?prid=5652759&bid=18508764)

**Comment by velddev** [comment]
Could you also show the new embeds in an image? That would be nice :D

File: src/Miki/AppProps.cs [hunk 1]
```diff
@@ -17,6 +17,26 @@ public static class Emoji
             public static string Mekos => "<:mekos:421972155484471296>";
             public static string Reddit => "<:reddit:704006214694076427>";
             public static string WeeklyEmbedIcon => "🏠";
+
+            public const string Ledger = "📒";
+            public const string Running = "🏃‍";
+            public const string HammerAndWrench = "🛠";
+            public const string TV = "📺";
```
Comment (velddev) [inline src/Miki/AppProps.cs]: nit: Would prefer `Television` personally, but that's up to you

File: src/Miki/MikiBotApp.cs [hunk 2]
```diff
@@ -315,6 +315,9 @@ public override async Task ConfigureAsync(ServiceCollection serviceCollection)
             serviceCollection.AddSingleton(
                 x => new CommandTreeBuilder(x).Create(Assembly.GetExecutingAssembly()));
 
+            serviceCollection.AddSingleton(
+                x => new CommandTreeService(x.GetService<CommandTree>()));
```
Comment (velddev) [inline src/Miki/MikiBotApp.cs]: could be simplified by doing ```serviceCollection.AddSingleton<CommandTreeService>()```

File: src/Miki/Modules/GeneralModule.cs [hunk 3]
```diff
@@ -326,153 +327,196 @@ await builder.ToEmbed()
...
 		public async Task HelpAsync(IContext e)
-		{
-			var commandHandler = e.GetService<CommandTree>();
+        {
+            var commandHandler = e.GetService<CommandTreeService>();
             var locale = e.GetLocale();
+            var permissionService = e.GetService<PermissionService>();
+            var scopesService = e.GetService<ScopeService>();
+
+            var embedBuilder = new EmbedBuilder
```
Comment (velddev) [inline src/Miki/Modules/GeneralModule.cs]: We've been trying to keep new EmbedBuilder().SetColor(...)

**Comment by exsersewo** [comment]
> Could you also show the new embeds in an image? That would be nice :D

Here you go
image
image
image

**Comment by velddev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 19
           

Complexity increasing per file
==============================
- src/Miki/Services/CommandTreeService.cs  2
- src/Miki/Attributes/EmojiAttribute.cs  1
- src/Miki/Modules/GeneralModule.cs  1
         

Clones added
============
- src/Miki/Services/CommandTreeService.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/velddev/Miki/pullRequest?prid=5652759&bid=18508764)

File: src/Miki/Modules/GeneralModule.cs [hunk 4]
```diff
@@ -326,153 +327,196 @@ await builder.ToEmbed()
...
-
-            await new EmbedBuilder()
-			{
-				Description = e.GetLocale().GetString("miki_module_general_help_dm"),
-				Color = new Color(0.6f, 0.6f, 1.0f)
-			}.ToEmbed()
-				.QueueAsync(e, e.GetChannel())
+            else
+            {
+                await new EmbedBuilder()
```
Comment (velddev) [inline src/Miki/Modules/GeneralModule.cs:418]: Same here, would be good to try and keep embed builders consistent

File: src/Miki/Services/CommandTreeService.cs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                    .OfType<NodeModule>()
+                    .FirstOrDefault(x => x.Children
+                        .Any(z => z.Metadata.Identifiers
+                            .Any(y=>y.ToLowerInvariant() == commandName)
+                        )
+                    );
+
+            if(moduleNode == null)
+            {
+                return null;
```
Comment (velddev) [inline src/Miki/Services/CommandTreeService.cs]: perhaps a good idea to `throw EntityNullException<Node>("entity_command");` to make sure the return type is localized to the user.

File: src/Miki/Services/CommandTreeService.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            moduleName = moduleName.ToLowerInvariant();
+
+            var moduleNode = commandTree.Root.Children
+                    .OfType<NodeModule>()
+                    .FirstOrDefault(x => x.Metadata.Identifiers
+                        .Any(z => z.ToLowerInvariant() == moduleName));
+
+            if (moduleNode == null)
+            {
+                return null;
```
Comment (velddev) [inline src/Miki/Services/CommandTreeService.cs]: perhaps a good idea to throw EntityNullException<Node>("entity_module"); to make sure the return type is localized to the user.

File: src/Miki/Services/CommandTreeService.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            commandName = commandName.ToLowerInvariant();
+
+            var moduleNode = commandTree.Root.Children
+                    .OfType<NodeModule>()
+                    .FirstOrDefault(x => x.Metadata.Identifiers
+                        .Any(z=>z.ToLowerInvariant() == moduleName));
+
+            if (moduleNode == null)
+            {
+                return null;
```
Comment (velddev) [inline src/Miki/Services/CommandTreeService.cs]: perhaps a good idea to throw EntityNullException<Node>("entity_command"); to make sure the return type is localized to the user.

File: src/Miki/AppProps.cs [hunk 8]
```diff
@@ -17,6 +17,26 @@ public static class Emoji
             public static string Mekos => "<:mekos:421972155484471296>";
             public static string Reddit => "<:reddit:704006214694076427>";
             public static string WeeklyEmbedIcon => "🏠";
```
Comment (velddev) [inline src/Miki/AppProps.cs]: minor addition: to keep things consistent could you make these const too?

**Review by velddev — APPROVED** [review]
Seems good, I recommended a few changes to keep the code on the help clean.

**Comment by velddev** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 30
           

Complexity increasing per file
==============================
- src/Miki/Attributes/EmojiAttribute.cs  1
- src/Miki/Services/CommandTreeService.cs  2
- src/Miki/Modules/GeneralModule.cs  4
         

Clones added
============
- src/Miki/Services/CommandTreeService.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/velddev/Miki/pullRequest?prid=5652759&bid=18508764)

**Pull request merged**